src/generate-types.ts

```diff
@@ -1294,10 +1304,134 @@ export function run(options: Options) {
       ),
       generateTypeShape(key),
       generateTopLevelMaker(key, schema),
-      generateNamedTypeDefinitionAssignment(key, schema)
+      generateNamedTypeDefinitionAssignment(key, schema, isA)
     ];
   }
 
+  function generateIsAForScalar(key: string) {
+    return ts.createArrowFunction(
+      undefined,
+      undefined,
+      [makeAnyProperty('value')],
+      undefined,
+      ts.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
+      ts.createCall(
+        ts.createPropertyAccess(
+          ts.createCall(ts.createIdentifier('make' + oautil.typenamify(key)), undefined, [
+            ts.createIdentifier('value')
+          ]),
+          'isSuccess'
+        ),
+        undefined,
+        []
+      )
+    );
+  }
+
+  function generateIsAForOneOf(oneOf: (oas.SchemaObject | oas.ReferenceObject)[]) {
+    const expression = getExpressionsForOneOfIsA(oneOf);
+    if (!expression) return;
+
+    return ts.createArrowFunction(
+      undefined,
+      undefined,
+      [makeAnyProperty('value')],
+      undefined,
+      ts.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
+      expression
+    );
+  }
+
+  function getExpressionsForOneOfIsA(
+    oneOf: (oas.SchemaObject | oas.ReferenceObject)[],
+    param = 'value'
+  ) {
+    const calls = oneOf.map(obj => {
+      if (oautil.isReferenceObject(obj)) {
+        return generateIsACall('type' + resolveRefToTypeName(obj.$ref).member, param);
+      } else if (obj.items) {
+        const arrayIsA = getExpressionsForArrayIsA(obj.items);
+        return arrayIsA ? ts.createParen(arrayIsA) : undefined;
+      }
+    });
+
+    // TODO support non-ref options in oneOf
+    if (calls.some(call => !call)) return;
+    assert(calls.length > 0, 'empty oneOf should not be possible');
+
+    return calls.length === 1
+      ? calls[0]
+      : calls
+          .slice(2)
+          .reduce(
+            (acc, memo) => ts.createBinary(acc, ts.SyntaxKind.BarBarToken, memo!),
+            ts.createBinary(calls[0]!, ts.SyntaxKind.BarBarToken, calls[1]!)
+          );
```

Comment:
ugly but I couldn't find any way to create `(a || b || c ...)` for as many values as I wanted